File: Orchestrators/DotNet/Context/AgentContext.cs
namespace AITrove.Context;

/// <summary>
/// Immutable per-run context. Agents read from it; only the orchestrator writes results into it.
/// </summary>
public sealed class AgentContext
{
    public required string ApiKey       { get; init; }
    public required string UserMessage  { get; init; }
    public required string SystemPrompt { get; init; }

    /// <summary>
    /// Freeform metadata agents can query.
    /// Convention: use lowercase snake_case keys, e.g. "code", "language", "pr_diff".
    /// </summary>
    public Dictionary<string, object> Metadata { get; } = [];

    /// <summary>Accumulated outputs from agents that have already run in this workflow.</summary>
    private readonly List<(string Name, string Output)> _results = [];

    public void AddAgentResult(string name, string output) =>
        _results.Add((name, output));

    public IReadOnlyList<(string Name, string Output)> AgentResults =>
        _results.AsReadOnly();

    /// <summary>
    /// Builds the synthesis prompt by prepending all agent outputs as labeled
    /// sections before the original user message. Fed to Claude for final synthesis.
    /// </summary>
    public string BuildSynthesisPrompt()
    {
        var sb = new System.Text.StringBuilder();

        foreach (var (name, output) in _results)
        {
            sb.AppendLine($"### {name} findings");
            sb.AppendLine(output);
            sb.AppendLine();
        }

        sb.AppendLine("### User request");
        sb.Append(UserMessage);
        return sb.ToString();
    }
}

File: Orchestrators/Python/aitrove_python/agent_context.py
from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(slots=True)
class AgentContext:
    api_key: str
    user_message: str
    system_prompt: str
    metadata: dict[str, str] = field(default_factory=dict)
    agent_results: list[tuple[str, str]] = field(default_factory=list)

    def add_agent_result(self, name: str, output: str) -> None:
        self.agent_results.append((name, output))

    def build_synthesis_prompt(self) -> str:
        parts: list[str] = []

        for name, output in self.agent_results:
            parts.append(f"### {name} findings")
            parts.append(output)
            parts.append("")

        parts.append("### User request")
        parts.append(self.user_message)
        return "\n".join(parts)
